proposals/distribution.md

```diff
@@ -0,0 +1,156 @@
+* “Creating Reference spec for optional DNS based naming & distribution”.
+    This entry conflates naming and distribution, which will be separated by this proposal.
+* “Standardizing on a particular Distribution method”.
+    This proposal will provide one (of possibly many) distribution specifications, so the old “There is no current agreement on how to distribute content” no longer applies.
+
+The following entries should be added to the [scope table][scope]:
+
+* “Specifying authentication and authorization schemes”.
+    Docker's current registry uses an [extension][token] of the [`Bearer`][rfc6750] [auth scheme][rfc7235-s2.1].
+    Work on specifying Docker's scheme will continue independently, and is orthogonal to the registry API.
+
+    * What: Specifying authentication and authorization schemes
+    * In/Out/Future: Out of scope
+    * Status: N/A
+    * Description: Defining protocols for authenticating and authorizing distribution access.
+    * Why: As an HTTP-based protocol, clients and servers can negotiate authentication via HTTP's [challenge-response authentication framework][rfc7235-s2.1].
+        There is no need for the distribution specification to choose a particular authentication scheme, because clients receiving 401 and 407 responses can use IANA's [HTTP Authentication Scheme Registry][iana-auth] to look up referenced schemes and take appropriate action.
+
+* “Creating a reference spec for optional DNS based naming and discovery”.
+    Discovery and registry protocols are completely separate and do not need to be added together.
+    This entry replaces part of the previous “Creating Reference spec for optional DNS based naming & distribution” entry.
+
+    * What: Creating a reference spec for optional DNS based naming and discovery
+    * In/Out/Future: In scope for future specification
+    * Status: Work not yet started
+    * Description: Define a protocol for resolving an image name to retrieval information.
+        When we address this, we will also allow for alternative name-to-image discovery protocols in parallel with the OCI-specified protocol.
+    * Why: It is reasonable to provide a standardized way to use DNS based distribution in conjunction with OCI without requiring its use.
+        There are many good use cases for DNS based distribution, but not all use cases support this.
+        Furthermore, encoding the location of a bundle into the bundle can cause issues with downloads from alternate locations other than the origin specified in the name.
+
+* “Specifying a distribution method”.
+    This entry replaces part of the previous “Creating Reference spec for optional DNS based naming & distribution” and “Standardizing on a particular Distribution method” entries.
+
+    Retrieving image indexes covers the current “tag listing” (e.g. “what named manifests are in `library/busybox`?”), because tags are entries in the image format's [`manifests` array][manifests].
```

Comment:
"because tags are entries in the image format's [`manifests` array][manifests]."

This statement is completely false.

Reply:
This is only true for image layout and has nothing to do with images distributed over this API.

Reply:
> "because tags are entries in the image format…"
>
> This statement is completely false.

Would you accept:

```markdown
… because tags can be represented as entries in the image format's [manifests array][manifests].
```

?

Reply:
No, because these are not true statements.

Reply:
> No, because these are not true statements.

[Image-spec says][1]:

> A common use case of descriptors with a "org.opencontainers.image.ref.name" annotation is representing a "tag" for a container image.

Can you go into more detail about how using `manifests` with `org.opencontainers.image.ref.name` annotations is not capable of representing image tags?

[1]: https://github.com/opencontainers/image-spec/blame/v1.0.1/image-layout.md#L154

Reply:
Yes, the addition of `ref.name` was a poorly thought addition and doesn't have a relationship here. They are separate things. Yes, they could represent image tags, but the tagging in distribution spec is correctly unique (ie points at a manifest list), whereas `ref.name` tries to embed something. As I've said about a million times, `ref.name` is only valid on the layout `index.json`. If we want to use here as a future change, the maintainers can make that decision. Either way, this statement is wholly confusing.

Reply:
> As I've said about a million times, `ref.name` is only valid on the layout `index.json`.

In that case, I don't see why tags belong in the distribution spec at all.  If [manifests][1] are the highest-level object that is represented in the API, then why bother with tag-listing at all?  Just require consumers to ask for exactly the manifest they need with something like:

    $ curl -H 'Accept: application/vnd.oci.image.manifest.v1+json' /v2/library/alpine/3.7

if, instead, you want to continue to provide [an endpoint that lists multiple tags][2], is there really a benefit to returning a list of strings instead of a [list of descriptors][3]?


[1]: https://github.com/opencontainers/image-spec/blob/v1.0.1/manifest.md
[2]: https://github.com/docker/distribution/blob/5cb406d511b7b9163bff9b6439072e4892e5ae3b/docs/spec/api.md#listing-image-tags
[3]: https://github.com/opencontainers/image-spec/blame/v1.0.1/image-index.md#L23

Reply:
> In that case, I don't see why tags belong in the distribution spec at all

That is something we could consider for enhancing the protocol. Right now the API can mostly be used with tags but manifests cannot be uploaded without a tag. It is best to think of manifests as the highest level object and tags as part of the API for usability. Tags separated from the API is sometimes useful for security, this is what notary was created to solve.

As for this API, can you please remove language linking tags with the manifests and indexes, that is confusing as there is no relationship within the API. You can keep language about tag listing endpoints as being in scope, although as mentioned, there are other ways to get lists of tags from external indexes.